Add MemFile text interface factored so users can optionally skip Nim …

c-blake opened the pull request:
…string

construction (e.g. to filter many lines out before building strings).  Only
Windows and/or Unix line endings are supported by default, though fiddling
with delim/eat arguments can probably support archaic MacOS9 files.

[ There is no SSE-optimized "mempbrk/memcspn" analogue to strpbrk or strcspn
 to make supporting all line ending styles both easy and super efficient. ]

In my timings, string construction is 2/3 of a trivial test program which means the
factoring allows some things to be 3X faster.

Some existing type may be more desirable than the local Record type.

Varriount commented on lib/pure/memfiles.nim:
```diff
@@ -245,3 +245,35 @@ proc close*(f: var MemFile) =
+iterator records*(mfile: MemFile, delim='\l', eat='\r'): Record {.inline.} =
+  proc c_memchr(cstr: cstring, c: char, n: csize): cstring {.
+       importc: "memchr", header: "<string.h>" .}
+  proc `-!`(p, q: cstring): int {.inline.} = return cast[int](p) -% cast[int](q)
```
You could just make this a template.

c-blake replied:
Which..? The whole quoted snippet or the `-!` ?

Varriount commented on lib/pure/memfiles.nim:
```diff
@@ -245,3 +245,35 @@ proc close*(f: var MemFile) =
   
   if error: raiseOSError(lastErr)
 
+type Record* {.unchecked.} = object
```
Lower case members here please. I would also prefer some documentation as to what the record type's purpose is. Should it be mutable? Immutable? (Really, I would prefer that the stdlib just get an immutable string type, then that could be used)

c-blake replied:
Easy re: the member spelling.  I honestly figured somewhere there was probably already some kind of type better than Record, maybe some kind of Slice or precursor to it, and thought of Record as a placeholder.

As to purpose, it is only to abstract a slice but in terms of a Cstring buffer - both so one could easily filter lines by saying, e.g. 'if rec.Len >3 and memcmp(rec.Beg, "blah", 4): ...' and also to be clear that the memory is unmanaged unlike most other Nim situations.  Fine to add docs, but thought it wise to first get feedback on the type itself.

As to mutability...Well, that is important, but it also "just depends".  If the MemFile is writable then so are the virtual memory image slices returned by the iterator.  Also, that decision is made at a higher level of context -- by the builder of the MemFile.  While 95+% of use cases might be read-only, read-write ones that only modify the file "in-place" are imaginable.  It also bears noting that the lifetime of the pointed-to memory is also managed here by callers as well as the permission.  So, there is implicitly very much an assumption of people keeping track of what constraints exist on the memory.  Factoring the API differently would probably be desired to reflect both kinds of mutability in the types involved.

Araq replied:
I would call it ``MemSlice`` instead of ``Record`` I think.

Varriount commented:
Overall, this is pretty good, however I would like to think on if the code should be merged right away. The general topic of 'views' on containers such as strings, sequences, etc is something that should probably be investigated a bit first.
Put simply: I would dislike it if this addition was merged, only to have a similar (but slightly different) record/view object or interface be created in another part of the standard library.

c-blake commented:
Well, I don't know what views plans might be in flight.  It's a pretty small little nugget of functionality that ties in well with MemFile, though.  I doubt it would be all that hard to re-do the return type in terms of some future views style.

I considered using MemFile itself as the record type since such records are basically "sub-files" and in this particular instance the type itself is enough to support a view on itself.  I thought the unused file handles (non-exported thought they are) might be more confusing than that was worth, though...maybe a kind of clarity/purity trade-off.  Also, cstring is a tiny bit nicer than pointer for using the buffer.

I might say to not export the data members to help such, but they are actually needed by the caller to do any pre-Nim-stringification..exported cstring & len is sort of the whole point of Record.

Araq commented on lib/pure/memfiles.nim:
```diff
@@ -245,3 +245,35 @@ proc close*(f: var MemFile) =
+proc toString*(rec: Record): string {.inline.} =
+  proc toNimStr(str: cstring, len: int): string {. importc: "toNimStr" .}
+  result = toNimStr(cast[cstring](rec.data), rec.size)
+  result[result.len] = '\0'
```
Not required, toNimStr should do the trailing binary zero already.

Araq commented on lib/pure/memfiles.nim:
```diff
@@ -245,3 +245,35 @@ proc close*(f: var MemFile) =
+       importc: "memchr", header: "<string.h>" .}
+  proc `-!`(p, q: cstring): int {.inline.} = return cast[int](p) -% cast[int](q)
+  var rec: Record
+  var End: cstring
```
I usually use 'last' instead of 'end' or perhaps 'ending'. Upper cased just because end is a keyword already is kind of ugly.

c-blake commented:
I made the identifier changes you suggested (and one more for the local 'rec' -> ms for an (M)em(S)lice instance).

While your intent for toNimStr may be to ensure a NUL terminator, I do not believe it does that right now.  It does allocate and memcpy an extra byte to copy _an existing_ NUL byte.  In this use case, though, the file has non-NUL bytes past the end.  So, the memcpy will just copy whatever byte is there, usually a delimiter or fragment of a delimiter (e..g, \r or \l).  I fully support some kind of toNimStr guarantee of C termination and would gladly delete that line in this PR when it's there.  I see two ways - adding a line like my foo[foo.len] = 0 after the memcpy or alternately, memcpy one fewer bytes if the rawNewStringNoInit freshly allocated memory could be assume zeroed.

On a termination related note, I also wasn't sure if 'cstring' was the least confusing type for the MemSlice.data pointer to non-terminated character array.  Users will probably mostly be using C 'mem' functions on these objects, not 'str' functions.  In C those usually take "void*".  So, maybe we should use 'pointer'?  I'm happy to also make that edit for this PR.

c-blake commented:
As an additional observation/bug report on toNimStr(), it does access input memory past the end of the caller's argument assuming that it will be a valid memory reference (as well as being NUL).  Being only one byte, it's a rare case to be sure, but it is certainly an unusual/counter-intuitive API that the caller must ensure an extra byte can be read than passed to the callee.  It's not easy/possible at all to do that with an memory map.  The map case would likely only bite you for a file that was a multiple of 4096 bytes (or whatever the page size is on the platform) since access permissions tend to be "by page".

Anyway, probably it would be best to chang toNimStr to be something like (dropping the +1 in memcpy and adding a line):
```
proc toNimStr(str: cstring, len: int): NimString {.compilerProc.} =
  result = rawNewStringNoInit(len)
  result.len = len
  c_memcpy(result.data, str, len)
  result.data[len] = '\0'
```

c-blake commented:
Well, I just made that change I suggested for concretneess.  Testing-wise, the toNimStr fix obviously impacts a lot more code than just the off to the side addition to memfiles.nim.  That said, it tests out ok on my system and seems to follow your original intent and also won't segfault on rare input like the prior impl as per my note above.

c-blake commented:
If for some reason you happen to be against making sysstr.nim:toNimStr robust to unterminated inputs or inputs with a final byte past readable memory (i.e. not C string inputs) then in this PR we could go back to doing the terminating inside toString().  However, in that case the toString() of this PR would need a bit more logic to ensure the old memcpy(..., len+1) in toNimStr() doesn't read past the end of valid memory (passing len - 1, maybe a zero len case, etc.).

Araq commented:
> If for some reason you happen to be against making sysstr.nim:toNimStr robust to unterminated inputs

No, I'm not, I'm just overwhelmed with work, ok? ;-)

c-blake commented:
Ok.  No problem at all.  :-)  From my perspective I can't think of any other ways to improve this PR.  Let me know if you want anything more.  Good luck catching up.

Araq commented:
Requires tests. Fixing toNimStr could be a separate PR. 'unchecked' for object has no semantics.

c-blake commented:
Done Re: unchecked.

There is no test for any of memfiles.nim.   I guess that's what you want me to write..a "tests/misc/memfiles.nim"..and maybe use that "readme.txt" for input?

It might also be good to add another, faster mode of use (say "memfiles.readLine") where a single string object can be re-used as the output buffer.

c-blake commented:
Ok.  I did a few simple test/demo programs.  I also re-organized the procs and I added some documentation comments and examples and removed use of toNimStr().  { Once this PR is settled I'll make another super simple 2-liner one to fix toNimStr if you want. }

The new version has a more efficient lines() iterator anyway, only just about 1.4X more expensive than iterating over memSlices().  It also uses TaintedString to be consistent with the readLine/lines stuff.

dom96 commented on lib/pure/memfiles.nim:
```diff
@@ -245,3 +248,73 @@ proc close*(f: var MemFile) =
+
+proc c_memcpy(a, b: pointer, n: int) {.importc: "memcpy", header: "<string.h>".}
+
+proc toString*(ms: MemSlice): string {.inline.} =
```
This should be defined as `$`.

dom96 commented:
I would also like to see the documentation improved a bit, it's currently a bit rough and contains some grammatical errors.

dom96 commented:
The tests should be in the ``stdlib`` category.

c-blake commented:
So..I'm not sure the things I tweaked were the grammatical errors you were thinking about.  I'm happy to either expand some stuff or make other changes, but it's hard to respond to "rough".  E.g., do you think there should be a big bold flag that "\\r only"-delimited archaic MacOS 9 files aren't covered by the default?  They are covered with non-default parameters, of course.

Also, feel free to just tweak the doc comments yourself post-merge!  I personally find that documentation updates are often best driven iteratively by end-user questions.

BTW, currently, the build is broken (lib/core/macros.nim fails) which slowed me down a smidge.

dom96 commented:
Reads much better now than it used to :)

Macros.nim building should be fixed now.

c-blake commented:
Ok.  Good to hear.  Let me know if you want/think of any other changes.  Building works again for me.  Thanks.
